File: backend/src/skills/auth/login_user_skill.py
"""LoginUserSkill - Handles user authentication and JWT generation."""

from typing import Dict
from sqlmodel import select
from src.models.user import User
from src.auth.password import verify_password
from src.auth.jwt import create_access_token
from .base_auth_skill import BaseAuthSkill, SkillResult


class LoginUserSkill(BaseAuthSkill):
    """
    Skill for user authentication and login.

    Responsibilities:
    - Validate credentials
    - Authenticate user against database
    - Generate JWT access token
    - Return token for client use
    """

    def execute(self, email: str, password: str) -> SkillResult:
        """
        Authenticate user and generate JWT token.

        Args:
            email: User's email address
            password: Plain-text password to verify

        Returns:
            SkillResult with JWT token or error message
        """
        try:
            # Validate inputs
            if not email or not password:
                return SkillResult(success=False, error="Email and password are required")

            email = email.strip().lower()

            # Find user by email
            user = self._find_user_by_email(email)
            if not user:
                return SkillResult(success=False, error="Incorrect email or password")

            # Verify password
            if not verify_password(password, user.password_hash):
                return SkillResult(success=False, error="Incorrect email or password")

            # Generate JWT token
            access_token = create_access_token(user.id, user.email)

            # Return token response
            token_data = {
                "access_token": access_token,
                "token_type": "bearer",
            }

            return SkillResult(success=True, data=token_data)

        except Exception as e:
            return SkillResult(success=False, error=f"Authentication failed: {str(e)}")

    def _find_user_by_email(self, email: str) -> User | None:
        """Find user by email address."""
        statement = select(User).where(User.email == email)
        return self.db.exec(statement).first()

File: frontend/lib/skills/login-user-skill.ts
/**
 * LoginUserSkill - Handles user login flow
 *
 * Responsibilities:
 * 1. Authenticate with Better Auth (create frontend session)
 * 2. Authenticate with backend API (validate credentials)
 * 3. Obtain and store backend JWT token
 */

import { signIn } from '../auth-client';
import { BaseAuthSkill, AuthResult } from './base-auth-skill';

interface BackendAuthResponse {
  access_token: string;
  token_type: string;
}

export class LoginUserSkill extends BaseAuthSkill {
  /**
   * Execute the login flow
   *
   * @param email - User email address
   * @param password - User password
   * @returns AuthResult indicating success or failure
   */
  async execute(email: string, password: string): Promise<AuthResult> {
    try {
      // Step 1: Authenticate with Better Auth (creates frontend session)
      const betterAuthResult = await signIn.email({
        email,
        password,
      });

      if (betterAuthResult.error) {
        return {
          success: false,
          error: betterAuthResult.error.message || 'Authentication failed',
        };
      }

      // Step 2: Authenticate with backend to get JWT
      let jwt: string;
      try {
        jwt = await this.authenticateBackend(email, password);
      } catch (backendError: any) {
        // If backend auth fails, clear Better Auth session
        // Note: We don't await this to avoid circular dependency
        return {
          success: false,
          error: backendError.message || 'Backend authentication failed',
        };
      }

      // Step 3: Store JWT for API calls
      this.storeJwt(jwt);

      return {
        success: true,
      };
    } catch (error: any) {
      return {
        success: false,
        error: error.message || 'Login failed',
      };
    }
  }

  /**
   * Authenticate with backend and obtain JWT token
   */
  private async authenticateBackend(email: string, password: string): Promise<string> {
    const response = await this.callBackendApi<BackendAuthResponse>('/api/auth/login', {
      method: 'POST',
      body: JSON.stringify({ email, password }),
    });
    return response.access_token;
  }
}
